### PR TITLE
Enable Flash and Fused attn with CP and THD

### PR DESCRIPTION
**Pull request by ipanfilo**
# Description

Fix and enable Fused/Flash attention with CP with THD format

Fixes [#10988](https://github.com/ROCm/frameworks-internal/issues/10988)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Don't fail FlashAttn tests if padded regions of output tensors are not zeroed
- Use correct cu_seqlen format for ROCm fused attn tests

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/fused_attn/test_fused_attn_with_cp.py [hunk 1]
```diff
@@ -122,9 +120,7 @@ def test_cp_with_flash_attention(dtype, model, qkv_format, cp_comm_type):
 @pytest.mark.parametrize("qkv_format", ["bshd", "sbhd", "thd"])
 @pytest.mark.parametrize("cp_comm_type", ["p2p", "all_gather", "a2a", "a2a+p2p"])
 def test_cp_with_fused_attention(dtype, model, qkv_format, cp_comm_type):
-    if IS_HIP_EXTENSION and qkv_format == "thd":
-        pytest.skip("THD format in CP has not been supported on ROCm yet!")
-    if qkv_format == "thd" and get_device_compute_capability() < (9, 0):
+    if (not IS_HIP_EXTENSION) and qkv_format == "thd" and get_device_compute_capability() < (9, 0):
```
Comment (wangye805) [inline tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: Can our fused attn backend support CP now? I recall there was a requirement that "pad_between_seq" need to be true to support fused attn in CP with THD. But I haven't enabled that...

Comment (wangye805) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: For example, can you check whether this pytest (tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[p2p-bshd-cp_2_3-bf16]) is actually using flash-attn or fused attn as backend or not? Though our CI shows it passed

Comment (ipanfilo) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: If we're talking about the same padding, it is rather opposite - for THD it asserts no padding: https://github.com/ROCm/TransformerEngine/blob/06ed6c9c1c1af39ade705eb66a93e62e8675a9ee/transformer_engine/pytorch/attention.py#L524 and change at https://github.com/ROCm/TransformerEngine/blob/489903345d84dac16b2d44df55a632a0f0f81e82/tests/pytorch/fused_attn/run_fused_attn_with_cp.py#L162 makes it passing

Comment (wangye805) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: Can you quickly run this tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[p2p-bshd-cp_2_3-bf16 and enable some logging variables like NVTE_LOG_FUSED_ATTN_CONFIG or NVTE_LOG_CK_CONFIG to ensure that indeed ck backend fused attn is running? Because I somehow remember last time when I try to run this but got blocked by this exact same line: https://github.com/ROCm/TransformerEngine/blob/06ed6c9c1c1af39ade705eb66a93e62e8675a9ee/transformer_engine/pytorch/attention.py#L524

Comment (ipanfilo) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: Why are we talking about BSHD at all?

Comment (wangye805) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: Sorry I copied the wrong test name. How about this one: [2025-03-07T03:06:44.686Z] ../tests/pytorch/fused_attn/test_fused_attn_with_cp.py::test_cp_with_fused_attention[p2p-thd-cp_1_0-bf16] PASSED [ 45%]
I suspect it may actually be running the flash-attn backend

Comment (ipanfilo) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: I validated CP tests with THD for both Flash and Fused attention.
One thing I didn't do, is removing check for padding attention mask from TE CK code and see if it works with causal mask too

Comment (wangye805) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: Cool. Thanks for the validation

**Review by wangye805 — APPROVED** [review]
LGTM

File: tests/pytorch/fused_attn/test_fused_attn_with_cp.py (hunk 1 above)
Comment (wenchenvincent) [reply tests/pytorch/fused_attn/test_fused_attn_with_cp.py:123]: `pad_between_seqs` is determined by `cu_seqlens_q_padded` and `cu_seqlens_q`
https://github.com/ROCm/TransformerEngine/blob/06ed6c9c1c1af39ade705eb66a93e62e8675a9ee/transformer_engine/pytorch/attention.py#L8117-L8123

so the change at
[TransformerEngine/tests/pytorch/fused_attn/run_fused_attn_with_cp.py](https://github.com/ROCm/TransformerEngine/blob/489903345d84dac16b2d44df55a632a0f0f81e82/tests/pytorch/fused_attn/run_fused_attn_with_cp.py#L162)

make `pad_between_seqs=False`

**Review by wenchenvincent — APPROVED** [review]
LGTM.
